OWML.ModLoader/ModSorter.cs

```diff
@@ -0,0 +1,78 @@
+﻿using OWML.Common;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OWML.ModLoader
+{
+    class ModSorter
+    {
+        public IList<ModDep> SortMods(IList<IModData> mods)
+        {
+            var modList = new List<ModDep>();
+            foreach (var mod in mods)
+            {
+                 modList.Add(new ModDep(mod.Manifest.UniqueName, mod, mod.Manifest.Dependencies));
+            }
+
+            return Sort(modList, x => x.Dependencies, x => x.Name);
+        }
+
+        private IList<T> Sort<T>(IEnumerable<T> source, Func<T, IEnumerable<T>> getDependencies)
+        {
+            var sorted = new List<T>();
+            var visited = new Dictionary<T, bool>();
+
+            foreach (var item in source)
+            {
+                Visit(item, getDependencies, sorted, visited);
+            }
+
+            return sorted;
+        }
+
+        private void Visit<T>(T item, Func<T, IEnumerable<T>> getDependencies, List<T> sorted, Dictionary<T, bool> visited)
+        {
+            var alreadyVisited = visited.TryGetValue(item, out bool inProcess);
+
+            if (alreadyVisited)
+            {
+                if (inProcess)
+                {
+                    throw new ArgumentException("Cyclic dependency found.");
+                }
+            }
+            else
+            {
+                visited[item] = true;
+
+                var dependencies = getDependencies(item);
```

Comment:
Trying to understand why you can't just do `var dependencies = item.Dependencies` here and get rid of this getDependencies argument completely. Same with getKey argument in other places. Is there some place I'm not seeing where you pass a different way of getting these attributes? These arguments are making this file very hard to follow and I don't see why.

Reply:
believe this is fixed in the rewrite